app/master/cluster_master.py

```diff
@@ -279,7 +282,7 @@ def handle_result_reported_from_slave(self, slave_url, build_id, subjob_id, payl
         build = self._all_builds_by_id[int(build_id)]
         slave = self._all_slaves_by_url[slave_url]
         # If the build has been canceled, don't work on the next subjob.
-        if not build.is_finished:  # WIP(joey): This check should be internal to the Build object.
+        if not build.is_finished or build.has_error:  # WIP(joey): This check should be internal to the Build object.
```

Comment:
I wasn't able to confirm tj's theory.

I launched a master and 2 slaves on my machine, both slaves had only 1 executor. I made a setup_build that immediately failed the first time it was run, but the second time it waited 5 seconds and then exited 0. If tj's theory was correct, then I would've expected slave 2 to not be freed, but it was.

I asked huckphin why he put this in there on the medium fork https://github.com/Medium/ClusterRunner/commit/fd38c99155eb0373815b75230f90f09295861770#commitcomment-22775741. I'm just gonna merge this tonight without the `or error`. If huckphin gives me a valid reason to leave this in after I merge, i'll do a follow up change.

https://github.com/Medium/ClusterRunner/commit/fd38c99155eb0373815b75230f90f09295861770#commitcomment-22775741